File: tests/room-brain-conformance.test.js
'use strict';

// RoomBrain conformance: Node leg.
//
// The same op log runs three times against the same module — here in Node, in
// JavaScriptCore on tvOS (RoomBrainConformanceTests.swift), and in QuickJS on
// Android TV (RoomBrainConformanceTest.kt) — and every step's return value and
// resulting snapshot must match tests/fixtures/room-brain-golden.json exactly.
//
// This leg has two jobs. It proves the golden still describes the module (so a
// behaviour change has to be an explicit regeneration, reviewed in the diff),
// and it proves the golden is reachable and well-formed for the two native legs
// that cannot run Node.

const { test, describe } = require('node:test');
const assert = require('node:assert/strict');
const fs = require('node:fs');
const path = require('node:path');
const { RoomBrain } = require('../server/RoomBrain.js');
const { INIT, OPS } = require('./fixtures/room-brain-ops.js');
const { replay } = require('../scripts/gen-room-brain-golden.js');

const GOLDEN_PATH = path.join(__dirname, 'fixtures', 'room-brain-golden.json');

describe('RoomBrain conformance', () => {
  test('the checked-in golden is present and matches the op log', () => {
    assert.ok(
      fs.existsSync(GOLDEN_PATH),
      'tests/fixtures/room-brain-golden.json is missing — run `npm run build:golden`'
    );
    const golden = JSON.parse(fs.readFileSync(GOLDEN_PATH, 'utf8'));
    assert.deepEqual(golden.init, INIT, 'golden init options are stale');
    assert.deepEqual(golden.ops, JSON.parse(JSON.stringify(OPS)), 'golden op log is stale');
  });

  test('replaying the op log reproduces the golden step for step', () => {
    const golden = JSON.parse(fs.readFileSync(GOLDEN_PATH, 'utf8'));
    const actual = replay();
    assert.equal(actual.length, golden.steps.length);
    for (let i = 0; i < actual.length; i++) {
      const label = OPS[i].g !== undefined
        ? `step ${i} (get ${OPS[i].g})`
        : `step ${i} (${OPS[i].m})`;
      assert.deepEqual(actual[i].result, golden.steps[i].result, `${label}: return value`);
      assert.deepEqual(actual[i].snapshot, golden.steps[i].snapshot, `${label}: snapshot`);
    }
  });

  test('the seeded rng makes auto-naming reproducible', () => {
    // Auto-naming is deliberately random, and was one of the behaviours that
    // had silently diverged across the three platforms. It has to be IN the
    // conformance log, so it has to be seedable across a JSON-only bridge.
    const names = () => {
      const b = new RoomBrain(INIT);
      b.peerJoined(1, 0); b.peerJoined(2, 0); b.peerJoined(3, 0);
      return [1, 2, 3].map((i) => b.get(i).playerName);
    };
    assert.deepEqual(names(), names());
    assert.ok(names().every((n) => /^HX-([1-9][0-9]?)$/.test(n)), 'auto names are HX-<n>');
    assert.equal(new Set(names()).size, 3, 'auto names are room-unique');
  });

  test('the blocklist is honoured on every platform, because there is only one', () => {
    // tvOS used to name players "HX-(slot + 1)" with no blocklist at all, so a
    // real Apple TV room could seat HX-4 and HX-13. The whole point of a shared
    // module is that this cannot be true on one platform and false on another.
    assert.deepEqual(RoomBrain.AUTO_NAME_BLOCKLIST, [4, 13, 17, 69]);
    const b = new RoomBrain({ rngSeed: 1 });
    for (let i = 0; i < 200; i++) {
      const name = b.generateAutoName(null);
      const num = Number(name.slice(3));
      assert.ok(!RoomBrain.AUTO_NAME_BLOCKLIST.includes(num), `generated a blocklisted ${name}`);
    }
  });
});

File: android/core/src/jvmTest/kotlin/com/hexstacker/core/room/RoomBrainConformanceTest.kt
package com.hexstacker.core.room

import com.hexstacker.core.engine.EngineBridge
import kotlinx.coroutines.runBlocking
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import java.io.File
import kotlin.test.Test
import kotlin.test.assertEquals

/**
 * RoomBrain conformance: the Android TV leg.
 *
 * `tests/fixtures/room-brain-golden.json` is a self-contained artifact — the
 * constructor options, an 88-step op log, and the expected return value plus full
 * snapshot after every step. Node replays it (tests/room-brain-conformance.test.js),
 * JavaScriptCore replays it on tvOS, and QuickJS replays it here.
 *
 * Because all three legs run the SAME JavaScript, this is not asking "did someone
 * re-implement the room wrong" any more. It asks "does this bridge marshal
 * correctly" — a far smaller surface, and the same shape as the engine's existing
 * FrameGoldenConformance gate. The op log deliberately covers the paths a
 * marshalling bug hides in: peer_joined before hello, colour collisions, auto-naming
 * with a blocklisted preference, name sanitizing (control chars, a zero-width
 * joiner, overlong, whitespace-only, legacy P-names), sparse AirConsole-style peer
 * indices, the three pause flags, suspend + cross-device rejoin, batched liveness
 * ticks, late-joiner grace, results enrichment, and a full room rejecting one more.
 */
class RoomBrainConformanceTest {

    private val json = Json { ignoreUnknownKeys = true }

    private fun bundle(): String =
        File(System.getProperty("hexcore.bundle") ?: error("hexcore.bundle not set by build")).readText()

    private fun golden(): JsonObject {
        val path = System.getProperty("hexcore.roombrain.golden")
            ?: error("hexcore.roombrain.golden not set by build")
        return json.parseToJsonElement(File(path).readText()).jsonObject
    }

    @Test
    fun replaysTheGoldenOpLogThroughTheBridge() = runBlocking {
        val g = golden()
        val ops = g["ops"]!!.jsonArray
        val steps = g["steps"]!!.jsonArray
        assertEquals(ops.size, steps.size, "golden ops/steps disagree")

        val bridge = EngineBridge.create(bundle())
        try {
            // rngSeed (not rng) rides the options object: the bridges are JSON-only and
            // cannot pass a function, and auto-naming — one of the behaviours that had
            // silently diverged across the three platforms — has to be IN the log.
            bridge.roomInit(g["init"]!!.jsonObject.toString())

            for (i in 0 until ops.size) {
                val op = ops[i].jsonObject
                val get = op["g"]?.jsonPrimitive?.content
                val method = op["m"]?.jsonPrimitive?.content
                val label = if (get != null) "step $i (get $get)" else "step $i ($method)"

                val returned = if (get != null) {
                    bridge.roomGetJson(get)
                } else {
                    val args = op["a"]?.jsonArray ?: JsonArray(emptyList())
                    bridge.roomCallJson(method!!, args.toString())
                }

                assertEquals(
                    steps[i].jsonObject["result"],
                    json.parseToJsonElement(returned),
                    "$label: return value",
                )
                assertEquals(
                    steps[i].jsonObject["snapshot"],
                    json.parseToJsonElement(bridge.roomSnapshotJson()),
                    "$label: snapshot",
                )
            }
        } finally {
            bridge.close()
        }
    }

    /** The window the shell's publish timer runs on comes from the module through the
     *  bridge, so the policy cannot be hand-mirrored into a Kotlin constant that drifts. */
    @Test
    fun exposesTheModuleSnapshotThrottle() = runBlocking {
        val bridge = EngineBridge.create(bundle())
        try {
            val client = RoomBrainClient.create(bridge, JsonObject(emptyMap()))
            assertEquals(500.0, client.snapshotThrottleMs())
        } finally {
            bridge.close()
        }
    }
}
